src/main/java/org/springframework/integration/kafka/core/DefaultConnectionFactory.java

```diff
@@ -172,13 +103,32 @@ public void setFetchMetadataTimeout(int fetchMetadataTimeout) {
 	 */
 	@Override
 	public BrokerAddress getLeader(Partition partition) {
+		BrokerAddress leader = null;
 		try {
 			lock.readLock().lock();
-			return this.getLeaders(Collections.singleton(partition)).get(partition);
+			if (getMetadataCache().containsPartition(partition)) {
+				leader = getMetadataCache().getLeader(partition);
+			}
 		}
 		finally {
 			lock.readLock().unlock();
 		}
+		if (leader == null) {
+			try {
+				lock.writeLock().lock();
+				this.refreshMetadata(Collections.singleton(partition.getTopic()));
```

Comment:
Yeah, we can do that. Again, based on use it's a very unlikely situation - but we can skip that refresh here.